firedrake/mg/ufl_utils.py

```diff
@@ -167,10 +169,35 @@ def coarsen_function(expr, self, coefficient_mapping=None):
     new = coefficient_mapping.get(expr)
     if new is None:
         V = expr.function_space()
-        manager = firedrake.dmhooks.get_transfer_manager(expr.function_space().dm)
+        dm = expr.function_space().dm
+        manager = firedrake.dmhooks.get_transfer_manager(dm)
         V = self(expr.function_space(), self)
         new = firedrake.Function(V, name="coarse_%s" % expr.name())
         manager.inject(expr, new)
+
+        expr._child = weakref.proxy(new)
+
+        def inject_on_restrict(fine, mrestrict, rscale, inject, coarse):
+            #print("Inside our restrict hook!!!")
+            finectx = firedrake.dmhooks.get_appctx(fine)
+
+            coeffs = finectx.F.coefficients() + finectx.J.coefficients()
+            if finectx.Jp is not None:
+                coeffs += finectx.Jp.coefficients()
+            coeffs = tuple(OrderedDict.fromkeys(coeffs))
+
+            for coeff in coeffs:
+                #print("Considering coeff = %s" % coeff)
+                if isinstance(coeff, firedrake.Function):
+                    if hasattr(coeff, '_child'):
+                        #print("Injecting %s -> %s" % (coeff, coeff._child))
+                        manager.inject(coeff, coeff._child)
```

Comment:
I think this function wants to be something like:

```
def inject_on_restrict(finedm, mrestrict, rscale, inject, coarse, fine=None, coarse=None):
    manager.inject(fine, coarse)

dm.addCoarsenHook(None, inject_on_restrict, kargs={"fine": expr, "coarse": expr._child})
```

We still need to handle the cleanup though.